webapp/src/idbworkspace.ts

```diff
@@ -17,47 +11,166 @@ interface StoredText {
+                continue;
+        }
 
-    await Promise.all(previousHeaders.map(h => copyProject(h)));
+        const db = await getDbAsync(prefix)
+        const existing = await db.getAsync(table, id);
+
+        if (!existing) {
+            await db.setAsync(table, entry);
+        }
+    }
+
+    await oldDb.deleteAllAsync(POUCH_OBJECT_STORE);
 }
 
-let _dbPromise: Promise<pxt.BrowserUtils.IDBWrapper>;
-async function getDbAsync(): Promise<pxt.BrowserUtils.IDBWrapper> {
-    if (_dbPromise) {
-        return await _dbPromise;
+async function migrateOldIndexedDb() {
+    const legacyDb = new pxt.BrowserUtils.IDBWrapper(`__pxt_idb_workspace`, 1, (ev, r) => {
+        const db = r.result as IDBDatabase;
+        db.createObjectStore(TEXTS_TABLE, { keyPath: KEYPATH });
+        db.createObjectStore(HEADERS_TABLE, { keyPath: KEYPATH });
+    }, async () => {
+        await pxt.BrowserUtils.clearTranslationDbAsync();
+        await pxt.BrowserUtils.clearTutorialInfoDbAsync();
+    });
+
+    try {
+        await legacyDb.openAsync();
+        const currentDb = await getCurrentDbAsync();
+
+        await copyTableEntries(legacyDb, currentDb, HEADERS_TABLE, true);
+        await copyTableEntries(legacyDb, currentDb, TEXTS_TABLE, true);
+    } catch (e) {
+        pxt.reportException(e);
     }
+}
+
+async function migratePrefixes() {
+    if (!getCurrentDBPrefix()) return;
+
+    const currentVersion = pxt.semver.parse(pxt.appTarget.versions.target);
+    const currentMajor = currentVersion.major;
+    const previousMajor = currentMajor - 1;
+    const previousDbPrefix = previousMajor < 0 ? "" : pxt.appTarget.appTheme.browserDbPrefixes[previousMajor];
+
+    if (!previousDbPrefix) return;
+    const currentDb = await getCurrentDbAsync();
 
-    _dbPromise = createDbAsync();
+    // If headers are already in the new db, migration must have already happened
+    if ((await currentDb.getAllAsync(HEADERS_TABLE)).length) return;
 
-    return _dbPromise;
+    const prevDb = await getDbAsync(previousDbPrefix);
+
+    await copyTableEntries(prevDb, currentDb, HEADERS_TABLE, false);
+    await copyTableEntries(prevDb, currentDb, TEXTS_TABLE, false);
+    await copyTableEntries(prevDb, currentDb, SCRIPT_TABLE, false);
+    await copyTableEntries(prevDb, currentDb, HOSTCACHE_TABLE, false);
+    await copyTableEntries(prevDb, currentDb, GITHUB_TABLE, false);
+}
+
+let _dbPromises: pxt.Map<Promise<pxt.BrowserUtils.IDBWrapper>> = {};
+
+async function getDbAsync(prefix = "__default") {
+    if (_dbPromises[prefix]) return _dbPromises[prefix];
+
+    _dbPromises[prefix] = createDbAsync();
+
+    return _dbPromises[prefix];
 
     async function createDbAsync(): Promise<pxt.BrowserUtils.IDBWrapper> {
-        const idbDb = new pxt.BrowserUtils.IDBWrapper("__pxt_idb_workspace", 1, (ev, r) => {
+        const idbDb = new pxt.BrowserUtils.IDBWrapper(`__pxt_idb_workspace_${pxt.storage.storageId()}_${prefix}`, 1, (ev, r) => {
```

Comment:
Just wondering, why was it necessary to append these things to the db name?

Reply:
the storage id is just for localhost so that different local served targets don't use the same db (we did the same thing for pouch).

the prefix is just to separate the different browser prefix versions of storage. in micro:bit we maintain separate dbs for different major versions of the editor (e.g. v0, v1, v2, etc.)